### PR TITLE
catalog: Add metrics to persist catalog

### LINKED ISSUE #23465: catalog: Add metrics to the persist catalog
### Feature request

The persist catalog doesn't contain any metrics tracking. Before enabling it in prod we should add metrics.

### PR DESCRIPTION
**Pull request by jkosh44**
Resolves #23465

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
I have fairly low confidence in code that I write related to metrics collection, so please let me know any improvements that you see.

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -533,6 +533,7 @@ impl Catalog {
             mz_catalog::durable::persist_backed_catalog_state(
                 persist_client,
                 environment_id.organization_id(),
+                &MetricsRegistry::new(),
```
Comment (danhhz) [inline src/adapter/src/catalog.rs:536]: personally, I'd make this a parameter and fill this `&MetricsRegistry::new()` in the single caller (testdrive) but ymmv

File: src/catalog/src/durable.rs [hunk 2]
```diff
@@ -284,7 +286,10 @@ pub async fn shadow_catalog_state(
     organization_id: Uuid,
 ) -> impl OpenableDurableCatalogState {
     let stash = Box::new(stash_backed_catalog_state(stash_config));
-    let persist = Box::new(persist_backed_catalog_state(persist_client, organization_id).await);
+    let persist = Box::new(
+        persist_backed_catalog_state(persist_client, organization_id, &MetricsRegistry::new())
```
Comment (danhhz) [inline src/catalog/src/durable.rs:290]: ditto here: I'd pass down the MetricsRegistry. in general, I try to only construct them in main files and things that are themselves obviously test code

File: src/catalog/src/durable/impls/persist/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+use mz_ore::metric;
+use mz_ore::metrics::{IntCounter, MetricsRegistry};
+use mz_ore::stats::histogram_seconds_buckets;
+use prometheus::{Histogram, IntCounterVec};
+
+#[derive(Debug, Clone)]
+pub struct Metrics {
+    pub transactions: IntCounter,
+    pub transaction_commit_errors: IntCounterVec,
+    pub transaction_commit_latency_duration_seconds: Histogram,
```
Comment (danhhz) [inline src/catalog/src/durable/impls/persist/metrics.rs:21]: Histograms are quite a bit more expensive than a counter, specifically they are equal to `N=2` counters where N is the number of buckets in the histogram. Plus, you can get pretty far by just using carefully selected normal counters. So, my rule of thumb is to never use a histogram until there's been a debugging incident that couldn't be diagnosed without the histogram. For reference, all of persist has 6 histograms (and I think one of them doesn't actually need to be a histogram)

Concretely, I think here you should start by replacing each of these histograms with one counter for the total number of seconds spend and one for the number of calls. That's enough to get you (1) number of calls (2) total time spent and (3) average time spent, which might very well be all you ever need

File: src/catalog/src/durable/impls/persist/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    pub snapshot_latency_duration_seconds: Histogram,
+    pub sync_latency_duration_seconds: Histogram,
+}
+
+impl Metrics {
+    /// Returns a new [Metrics] instance connected to the given registry.
+    pub fn new(registry: &MetricsRegistry) -> Self {
+        Self {
+            transactions: registry.register(metric!(name: "catalog_transactions", help: "Total number of started transactions.")),
```
Comment (danhhz) [inline src/catalog/src/durable/impls/persist/metrics.rs:30]: nit: I don't think rustfmt can understand macros, so it won't do this for you, but I'd put one thing per line here (like you have below) to make it a bit easier to scan

**Review by danhhz — COMMENTED** [review]
Left a few high level comments on the metrics in particular. You might also want to deploy this to your staging env and start a grafana dashboard to make sure that it feels like you're able to get the info you want

File: src/adapter/src/catalog.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/catalog.rs]: I think that sounds a lot cleaner, done.

File: src/catalog/src/durable.rs (hunk 2 above)
Comment (jkosh44) [reply src/catalog/src/durable.rs]: Done

File: src/catalog/src/durable/impls/persist/metrics.rs (hunk 4 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist/metrics.rs]: Done

File: src/catalog/src/durable/impls/persist/metrics.rs (hunk 3 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist/metrics.rs]: Done, let me know if the counter approach I used looks good.

File: src/catalog/src/durable/impls/persist/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                name: "catalog_transactions", 
+                help: "Total number of started transactions.",
+            )),
+            transaction_commit_errors: registry.register(metric!(
+                name: "catalog_transaction_errors", 
+                help: "Total number of transaction errors.", 
+                var_labels: ["cause"],
+            )),
+            transaction_commit_latency_duration_seconds: registry.register(metric!(
+                name: "catalog_transaction_latency",
```
Comment (mjibson) [inline src/catalog/src/durable/impls/persist/metrics.rs:42]:
```suggestion
                name: "catalog_transaction_latency_seconds",
```

https://prometheus.io/docs/practices/naming/#metric-names suggests that units should be in the name.

File: src/catalog/src/durable/impls/persist/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                help: "Latency for durable catalog transactions.",
+            )),
+            transactions_committed: registry.register(metric!(
+                name: "catalog_transactions_committed",
+                help: "Count of transactions committed.",
+            )),
+            snapshot_latency_duration_seconds: registry.register(metric!(
+                name: "catalog_snapshot_latency",
```
Comment (mjibson) [inline src/catalog/src/durable/impls/persist/metrics.rs:50]:
```suggestion
                name: "catalog_snapshot_latency_seconds",
```

File: src/catalog/src/durable/impls/persist/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                help: "Latency for fetching a snapshot of the durable catalog.",
+            )),
+            snapshots_taken: registry.register(metric!(
+                name: "catalog_snapshots_taken",
+                help: "Count of snapshots taken.",
+            )),
+            sync_latency_duration_seconds: registry.register(metric!(
+                name: "catalog_sync_latency",
```
Comment (mjibson) [inline src/catalog/src/durable/impls/persist/metrics.rs:58]:
```suggestion
                name: "catalog_sync_latency_seconds",
```

**Comment by jkosh44** [comment]
> You might also want to deploy this to your staging env and start a grafana dashboard to make sure that it feels like you're able to get the info you want

My plan is to deploy to staging once this PR and the following PRs are reviewed and merged:

- https://github.com/MaterializeInc/materialize/pull/23355
- https://github.com/MaterializeInc/materialize/pull/23388

They all make some significant changes to how the persist catalog is opened, read, and written. So it'd be nice to have them all deployed for long term staging testing.

**Review by mjibson — APPROVED** [review]

File: src/catalog/src/durable/impls/persist.rs [hunk 8]
```diff
@@ -828,31 +856,56 @@ impl DurableCatalogState for PersistCatalogState {
...
+        let res = commit_transaction_inner(self, txn_batch)
+            .wall_time()
+            .inc_by(counter)
+            .await;
+        self.metrics.transactions_committed.inc();
+        if let Err(e) = &res {
+            let cause = e.to_string_with_causes();
+            self.metrics
+                .transaction_commit_errors
+                .with_label_values(&[&cause])
```
Comment (danhhz) [inline src/catalog/src/durable/impls/persist.rs:905]: Any idea what the cardinality is here?

File: src/catalog/src/durable/impls/persist/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    pub sync_latency_duration_seconds: Counter,
+    pub syncs: IntCounter,
+}
+
+impl Metrics {
+    /// Returns a new [Metrics] instance connected to the given registry.
+    pub fn new(registry: &MetricsRegistry) -> Self {
+        Self {
+            transactions: registry.register(metric!(
+                name: "catalog_transactions",
```
Comment (danhhz) [inline src/catalog/src/durable/impls/persist/metrics.rs:33]: Ugh, hate to say this after you've merged, but I think we're pretty good about prefixing all the metrics in db code with `mz_`. Probably worth a followup, or just sneaking it into some other upcoming PR

File: src/catalog/src/durable/impls/persist.rs [hunk 10]
```diff
@@ -828,31 +856,56 @@ impl DurableCatalogState for PersistCatalogState {
...
+        }
+        let counter = self
+            .metrics
+            .transaction_commit_latency_duration_seconds
+            .clone();
+        let res = commit_transaction_inner(self, txn_batch)
+            .wall_time()
+            .inc_by(counter)
+            .await;
+        self.metrics.transactions_committed.inc();
```
Comment (danhhz) [inline src/catalog/src/durable/impls/persist.rs:900]: nit: is "committed" correct even if it errored?

**Review by danhhz — COMMENTED** [review]
New counters lgtm, thanks!

File: src/catalog/src/durable/impls/persist.rs (hunk 8 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist.rs:905]: Not really, is a high cardinality bad? If so, I think it's probably fine to remove the label.

File: src/catalog/src/durable/impls/persist/metrics.rs (hunk 9 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist/metrics.rs:33]: No worries, that's completely my fault for being impatient. I'll post a follow-up.

File: src/catalog/src/durable/impls/persist.rs (hunk 10 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist.rs:900]: Good point, I'll change the name and description to be more accurate.

File: src/catalog/src/durable/impls/persist.rs (hunk 8 above)
Comment (danhhz) [reply src/catalog/src/durable/impls/persist.rs:905]: Yeah, each unique value for the label ends up being a counter in prometheus

**Comment by jkosh44** [comment]
@danhhz Follow up is here: https://github.com/MaterializeInc/materialize/pull/23495

**Pull request merged**